client/hostvolumemanager/host_volumes.go

```diff
@@ -20,19 +22,59 @@ var (
 type HostVolumeManager struct {
 	pluginDir      string
 	sharedMountDir string
+	stateMgr       HostVolumeStateManager
 
 	log hclog.Logger
 }
 
-func NewHostVolumeManager(logger hclog.Logger, pluginDir, sharedMountDir string) *HostVolumeManager {
+func NewHostVolumeManager(logger hclog.Logger,
+	stateMgr HostVolumeStateManager, restoreTimeout time.Duration,
+	pluginDir, sharedMountDir string) (*HostVolumeManager, error) {
+
 	log := logger.Named("host_volume_mgr")
 
 	// db TODO(1.10.0): how do we define the external mounter plugins? plugin configs?
-	return &HostVolumeManager{
-		log:            log,
+	hvm := &HostVolumeManager{
 		pluginDir:      pluginDir,
 		sharedMountDir: sharedMountDir,
+		stateMgr:       stateMgr,
+		log:            log,
+	}
+
+	if err := hvm.restoreState(stateMgr, restoreTimeout); err != nil {
+		return nil, err
+	}
+
+	return hvm, nil
+}
+
+func (hvm *HostVolumeManager) restoreState(state HostVolumeStateManager, timeout time.Duration) error {
+	vols, err := state.GetDynamicHostVolumes()
+	if err != nil {
+		return err
+	}
+	if len(vols) == 0 {
+		return nil // nothing to do
 	}
+
+	// re-"create" the volumes - plugins have the best knowledge of their
+	// side effects, and they should be idempotent.
+	var wg sync.WaitGroup
+	for _, vol := range vols {
+		wg.Add(1)
+		func() {
```

Comment:
It looks like you meant for this to be run in a goroutine. This makes me realize another requirement we need to remember to document: plugins need to be safe to run concurrently.

Reply:
:+1: added a comment as part of error handling in 8043871641cb694d9c990d6a6a129508641591d3